processing/src/main/java/io/druid/segment/data/GenericIndexed.java

```diff
@@ -407,35 +216,56 @@ public T get(int index)
   @Override
   public int indexOf(T value)
   {
-    return bufferIndexed.indexOf(value);
+    return indexOf(this, value);
+  }
+
+  private int indexOf(Indexed<T> indexed, T value)
```

Comment:
> JIT won't necessarily figure out the right indexed. At least, if there are multiple concrete types of Indexed being passed, JIT might not be able to actually pick just a single one. Which will mean that every call to get will end up with an indirection. The previous implementation only has one indirection (the one for indexOf) and after that it's all local calls to the specific instance. I agree that this isn't the primary bottleneck, but am more wanting to make sure you understand that the previous implementation is better for JIT.

Before it was not a "direct" call also, it was a virtual call on unknown subclass of `BufferedIndexed`. So if `indexOf()` is not inlined as a whole, it was a normal `call` in assembly as well. The only difference between the former and the present version is that before JIT stored `this` pointer in a single register, and now it stores `this` and `indexed` separately (maybe in two different registers, or one of them on stack). This is minor.

> Given that I don't think that moving the indexOf is really buying us anything. It's either code in BufferedIndexed or it's up here. Why not just move it back to where it was and leave it be.

It couldn't be done because now `GenericIndexed.get()` doesn't use `bufferedIndexed` field, indirection removed. It's probably more important than the question with `indexOf()`, because `get()` is used much more actively than `indexOf()`